BHBC-2085

anissa-agahchen opened the pull request:
# Overview

## Links to Jira tickets
https://quartech.atlassian.net/browse/BHBC-2085

## Description of relevant changes
The work in this ticket is based on the assumption that the darwin_core_source column in the occurrence_submission is populated with the result of the transform as a JSON object.

- This code takes that object:
** finds all paths that contain a taxonID
** queries elastic search for the scientific and common name
** enriches the object at the same path with the scientific and common names
** returns the object

At a later point, this work will be merged with the transfom steps.

Note:  look at tests to see how the object transformations

anissa-agahchen committed:
getting the single species content from Elastic Search

anissa-agahchen committed:
storing scientific name in the incoming json object

anissa-agahchen committed:
can add scientific name as a child object, but not parents object

anissa-agahchen committed:
funny path problems

KjartanE committed:
Edits of jsonPath

anissa-agahchen committed:
merge conflict + sample code for json patch

anissa-agahchen committed:
object formatted correctly

anissa-agahchen committed:
some minor change

anissa-agahchen committed:
working code

anissa-agahchen committed:
minor change

anissa-agahchen committed:
clean up

anissa-agahchen committed:
remove taxon_authority, add english_name to output

anissa-agahchen committed:
wip - testing

anissa-agahchen committed:
working with new pattern

anissa-agahchen committed:
test for dwc-service

anissa-agahchen committed:
remove set up code

anissa-agahchen committed:
new test for taxonomy service

anissa-agahchen committed:
tweaks

anissa-agahchen committed:
clean up

anissa-agahchen committed:
test for multiple taxonIds at different depths

anissa-agahchen committed:
fixed patches

AlfredRosenthal commented on api/src/utils/media/xlsx/transformation/xlsx-transformation.ts:
```diff
@@ -58,7 +58,7 @@ export class XLSXTransformation {
 
   /**
-   * Flattens the worksheet data into arrays of objects.
+   * Flattens the worksheet data into arrays of objects.par
```
.par?

anissa-agahchen replied:
lol ... no idea what that's about

AlfredRosenthal commented on api/src/services/validation-service.ts:
```diff
@@ -135,6 +135,7 @@ export class ValidationService extends DBService {
       // template transformation
       await this.templateTransformation(submissionId, submissionPrep.xlsx, submissionPrep.s3InputKey, surveyId);
 
+      // insert template validated status
```
nit: next line is inserting transformed status?

anissa-agahchen committed:
clean ups

AlfredRosenthal reviewed (dismissed):
🐯

anissa-agahchen committed:
Merge branch 'dev' into BHBC-2085

anissa-agahchen committed:
Merge branch 'dev' into BHBC-2085

curtisupshall commented on api/src/services/dwc-service.ts:
```diff
@@ -0,0 +1,80 @@
+    this.xml2jsBuilder = new xml2js.Builder({ renderOpts: { pretty: false } });
+  }
+
+  async enrichTaxonIDs(jsonObject: Record<any, any>): Promise<Record<any, any>> {
```
Could we get some JS docs on this method?

curtisupshall commented on api/src/services/taxonomy-service.ts:
```diff
@@ -192,4 +192,47 @@ export class TaxonomyService {
     return response ? this._sanitizeSpeciesData(response.hits.hits) : [];
   }
+
+  private formatEnrichedData = (data: SearchHit<any>) => {
```
I believe making service methods private makes mocking the method more difficult or even impossible in the service tests. I think we should prefix this method with an `_` and make it public, like other methods in the class

curtisupshall replied:
Also, is it possible to be more specific with the typing on the `data` parameter? It should be `SearchHit<ITaxonomySource>` although I could be wrong

curtisupshall commented on api/src/services/taxonomy-service.ts:
```diff
@@ -192,4 +192,47 @@ export class TaxonomyService {
+    return { scientific_name, english_name };
+  };
+
+  async getEnrichedDataForSpeciesCode(code: string) {
```
Nit: JS docs for these two new methods

curtisupshall commented on api/src/services/dwc-service.test.ts:
```diff
@@ -0,0 +1,96 @@
+
+      expect(getEnrichedDataForSpeciesCodeStub).to.have.been.called;
+      expect(getEnrichedDataForSpeciesCodeStub).to.have.been.calledWith('M-OVCA');
+      expect(getEnrichedDataForSpeciesCodeStub).not.to.have.been.calledWith('M-OVCA1');
```
Nit: I think the above `expect` asserts that this will always be true

curtisupshall reviewed (dismissed):
I think it looks good! Just a handful of nitpicks re: comments and type defs

NickPhura committed:
Add JSDoc comments.
A few minor formatting tweaks.

NickPhura committed:
Merge remote-tracking branch 'remotes/origin/dev' into BHBC-2085

curtisupshall commented on app/src/features/surveys/view/SurveyObservations.tsx:
```diff
@@ -106,7 +106,7 @@ export enum SUBMISSION_MESSAGE_TYPE {
   'FAILED_PREP_XLSX' = 'Failed to prep XLSX',
   'FAILED_PERSIST_PARSE_ERRORS' = 'Failed to persist parse errors',
-  'FAILED_GET_VALIDATION_RULES' = 'Failed to get validation rules',
+  'FAILED_GET_VALIDATION_RULES' = 'Failed to get validation rules.  It is possible that your focal species is incorrect.',
```
I don't know how I feel about this addition. This enum is just supposed to store error message types; I think suggestions such as "fix your focal species" should belong in the actual error message, not in the message type string - especially given that the phrasing here is also a bit subjective. Moreover, if we did want to change this enum value, we would need a corresponding migration as well.

NickPhura replied:
Yeah makes sense.

I do think it adds additional information which is good, but it also could be misleading if the focal species isnt the issue (and you submitted the wrong template or something). The app side needs an update in general around how better to display error messages (rolling them up, having helper icons for more details, etc). So for now we can just leave it as it was, and a user may just need to reach out to Robin & Co for assistance (which they probably would need to anyways)

NickPhura committed:
Remove additional error message.
- It adds additional details, but may also lead to confusion if that taxon code isnt the issue.

curtisupshall reviewed (approved):
⛄
